File: solved/p303.py
import queue

def f(n):
    cnt = 0
    ck = [False]*n
    q = queue.Queue()
    q.put(1)
    q.put(2)
    while q.qsize() > 0:
        f = q.get()
        if ck[f%n] :
            continue
        ck[f%n] = True
        # print(cnt)
        if f%n == 0 :
            return f//n
        if not ck[(f*10)%n] :
            q.put(f*10)
        if not ck[(f*10+1)%n] :
            q.put(f*10 +1)
        if not ck[(f*10+2)%n] :
            q.put(f*10 +2)
    return 0

ans, SZ = 0, 10000

for i in range(1, SZ+1):
    ans += f(i)
    if i% 1000 == 0 : print(i)

print(ans)

# 4m 15s

File: solved/p303.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef pair<int, int> pii;
const ll SZ = 100;

ll process(ll i){
    queue<ll> q;
    vector<bool> ck(i,false);
    q.push(1ll); q.push(2ll);
    while(!q.empty()){
        ll f = q.front(); q.pop();
        if(ck[f%i]) continue;
        ck[f%i] = true;
        if(f%i==0) return f/i;
        q.push(f*10);
        q.push(f*10 + 1);
        q.push(f*10 + 2);
    }
    return 0;
}

int main(){
    ll ans = 1111333355557778;
    for(ll i = 1 ; i <= 9998 ; i++){
        ans += process(i);
    }
    ans += process(10000);
    cout << ans;
}
